### DISCUSSION TITLE
Add records to referenced sources

### DISCUSSION THREAD

**Discussion by JonathanThorpe** [savonet/liquidsoap#4067] [User Support]
Hi All,

I'm attempting to write a script that dynamically manages sources.

I've got this working:

```
...

def input_add(i) =
  log("Add stream #{i.id} / #{i.url}")
  stream = input.ffmpeg(i.url, id=i.id, start=false)
  streams.set(list.add(stream, streams()))
end

streams = ref([])
list.iter(input_add, config_data.inputs)

...

def ready_sources() = 
  list.find(default=fail, source.is_ready, streams())
end

radio = mksafe(source.dynamic(ready_sources))
```

In addition to this, I want to add some monitoring of the individual streams so that I can `source.start()` and `source.stop()` as required based on specific parameters that I manage through a separate thread.

I was hoping to add a record to the following:
```
stream = input.ffmpeg(i.url, id=i.id, start=false).{foo = false, bar = true}
```
And then dynamically set foo and bar later on, however I don't seem to be able to set these record attributes.

Is there a better approach?

**Comment by vitoyucepi** [comment]
Hi @JonathanThorpe,
The type of `config_data.inputs` is `[{url: string, id: string}]`
```ruby
def create_input(params)
  id = string.id.default(params.id, default="my_source")
  log("Add stream #{id} / #{params.url}")
  input.ffmpeg(params.url, id=id, start=false)
end

config_data = load_config()
streams = list.map(create_input, config_data.inputs)
```

But I don't understand this thing
```ruby
def ready_sources() = 
  list.find(default=fail, source.is_ready, streams())
end

radio = mksafe(source.dynamic(ready_sources))
```
Are you trying to find the first non-failed source? You may want to use `fallback(sources)` instead.

**Comment by JonathanThorpe** [comment]
Hi @vitoyucepi,

Regarding `radio = mksafe(source.dynamic(ready_sources))`, I incorrectly assumed that `fallback` requires sources to be defined as part of the configuration rather than runtime. Testing this however works, so I will change to that.

The problem I'm trying to solve is that I have a list of streams (ordered) and the highest priority stream is played unless a higher one becomes available at which time, switch to that. This is normal fallback behaviour but I don't want to run inactive streams which in some cases is running from a mobile connection. The streams are a combination of SRT (input.srt) and HTTP (input.ffmpeg).

I was thinking of tackling this by iterating through each stream and:
1. Starts the stream for 5 seconds to confirm it's alive.
2. Flag that the stream is active/inactive based on the above.

The above function would be initiated from a thread which runs every 5 minutes or triggered when the currently playing stream fails (perhaps `on_disconnect`).

To set the availability flag against the stream, I was looking to:
1. Define a ref that contains the streams:
streams = ref([])

2. Add each stream to `streams`.
3. Extend the record of each stream to contain the flag and then use the above-mentioned function to set it.

Perhaps my approach is over-complicating this?
